lib/spack/spack/compiler.py

```diff
@@ -549,24 +551,27 @@ def _compiler_environment(self):
         # store environment to replace later
         backup_env = os.environ.copy()
 
-        # load modules and set env variables
-        for module in self.modules:
-            # On cray, mic-knl module cannot be loaded without cce module
-            # See: https://github.com/spack/spack/issues/3153
-            if os.environ.get("CRAY_CPU_TARGET") == 'mic-knl':
-                spack.util.module_cmd.load_module('cce')
-            spack.util.module_cmd.load_module(module)
-
-        # apply other compiler environment changes
-        env = spack.util.environment.EnvironmentModifications()
-        env.extend(spack.schema.environment.parse(self.environment))
-        env.apply_modifications()
-
-        yield
-
-        # Restore environment
-        os.environ.clear()
-        os.environ.update(backup_env)
+        try:
+            # load modules and set env variables
+            for module in self.modules:
+                # On cray, mic-knl module cannot be loaded without cce module
+                # See: https://github.com/spack/spack/issues/3153
+                if os.environ.get("CRAY_CPU_TARGET") == 'mic-knl':
+                    spack.util.module_cmd.load_module('cce')
+                spack.util.module_cmd.load_module(module)
+
+            # apply other compiler environment changes
+            env = spack.util.environment.EnvironmentModifications()
+            env.extend(spack.schema.environment.parse(self.environment))
+            env.apply_modifications()
+
+            yield
+        except BaseException:
```

Comment:
LGTM.  Haven't run yet.   

Only suggestion -- since this is something I've noticed I need to work on -- is to provide a narrower set of exceptions here.

Reply:
+1 to unit test

For BaseException, I think it's appropriate here. The context manager relinquishes control via yield, so it should anticipate any error to be possible.

Reply:
Note I just ran the unit tests using `-k compiler` and encountered  failures and 1 error.